### PR TITLE
Gh 1847 create join op

### PR DESCRIPTION
**Pull request by m55624**

### CODE REVIEW

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    public ElementEquality() {
+    }
+
+    public ElementEquality(final Set<String> groupByProperties) {
+        this.groupByProperties = groupByProperties;
+    }
+
+    public ElementEquality(final String... groupByProperties) {
+        this.groupByProperties = Sets.newHashSet(groupByProperties);
```
Comment (gaffer01) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java]: Very minor, but should probably just add these properties to groupByProperties - a set is created at the top of the class so no need to create another one.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ */
+public class ElementEquality extends KoryphePredicate2<Element, Element> {
+    private Set<String> groupByProperties = new HashSet<>();
+
+
+    public ElementEquality() {
+    }
+
+    public ElementEquality(final Set<String> groupByProperties) {
+        this.groupByProperties = groupByProperties;
```
Comment (gaffer01) [inline core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java]: Maybe add all these to the existing set? It's safer as then the set won't change if the caller of this constructor changes the set.

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                .matchMethod(new TestMatch())
+                .matchKey(MatchKey.LEFT)
+                .joinType(JoinType.INNER)
+                .mergeMethod(new TestMerge())
+                .collectionLimit(10)
+                .build();
+    }
+
+    /**
+     * private copy of the ElementMatch class using the count property to match by.
```
Comment (gaffer01) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java]: This comment looks wrong.

File: core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    public static class TestMatch implements Match{
+        @Override
+        public List matching(final Object testObject, final List testList) {
+            return testList;
+        }
+    }
+
+    /**
+     * private copy of the ElementMatch class using the count property to match by.
```
Comment (gaffer01) [inline core/operation/src/test/java/uk/gov/gchq/gaffer/operation/impl/JoinTest.java]: Same comment as above.

File: core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java (hunk 1 above)
Comment (m55624) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java]: Do you mean do something like this? ```this.groupByProperties.addAll(Arrays.asList(groupByProperties));``` Is there any benefit to this?

Comment (gaffer01) [reply core/data/src/main/java/uk/gov/gchq/gaffer/data/element/comparison/ElementEquality.java]: Oh yes, you need to create a list before you can add it so not worth it - just thought creating another set was unnecessary.

**Pull request merged**